clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -19,8 +19,149 @@
+    } else if (_source_attribute !=
+               dpct::detail::pointer_access_attribute::host_only) {
+      target_t *host_ptr = dpct::get_host_ptr<target_t>(_source);
+      return sycl::buffer<target_t>(host_ptr, sycl::range<1>(_ele_num));
+    } else {
+      return sycl::buffer<target_t>(_source, sycl::range<1>(_ele_num));
+    }
+#else
+    if constexpr (std::is_same_v<target_t, source_t>) {
+      if (_source_attribute ==
+          dpct::detail::pointer_access_attribute::host_only)
+        return sycl::malloc_shared<target_t>(_ele_num, _q);
+      _need_free = false;
+      return _source;
+    } else {
+      return sycl::malloc_shared<target_t>(_ele_num, _q);
+    }
+#endif
+  }
+};
+
+template <typename target_t, typename source_t, mem_inout io>
+class mem_t : public mem_base_t<target_t, source_t> {
+  static_assert(io == mem_inout::out && "Only mem_inout::out is supported if "
+                                        "target_t and source_t are not same.");
+  using base_t = mem_base_t<target_t, source_t>;
+  using base_t::_q;
+  using base_t::_source;
+  using base_t::_source_attribute;
+  using base_t::_target;
+
+public:
+  mem_t(sycl::queue q, source_t *source) : base_t(q, source, 1) {}
+  ~mem_t() {
+#ifdef DPCT_USM_LEVEL_NONE
+    source_t temp = static_cast<source_t>(_target.get_host_access()[0]);
+    if (_source_attribute ==
+        dpct::detail::pointer_access_attribute::device_only) {
+      dpct::get_host_ptr<source_t>(_source)[0] = temp;
+    } else {
+      *_source = temp;
+    }
+#else
+    _q.wait();
+    source_t temp = static_cast<source_t>(*_target);
+    if (_source_attribute ==
+        dpct::detail::pointer_access_attribute::device_only) {
+      _q.memcpy(_source, &temp, sizeof(source_t)).wait();
+    } else {
+      *_source = temp;
+    }
+#endif
+  }
+};
+
+template <typename target_t, mem_inout io>
+class mem_t<target_t, target_t, io> : public mem_base_t<target_t, target_t> {
+  using base_t = mem_base_t<target_t, target_t>;
+  using base_t::_ele_num;
+  using base_t::_q;
+  using base_t::_source;
+  using base_t::_source_attribute;
+  using base_t::_target;
+  using typename base_t::data_t;
+
+public:
+  mem_t(sycl::queue q, target_t *source, size_t ele_num = 1)
+      : base_t(q, source, ele_num) {
+#ifndef DPCT_USM_LEVEL_NONE
+    if constexpr (io != mem_inout::out) {
+      if (_source_attribute ==
+          dpct::detail::pointer_access_attribute::host_only) {
+        _q.memcpy(_target, _source, sizeof(target_t) * _ele_num);
+      }
+    }
+#endif
+  }
+  ~mem_t() {
+#ifndef DPCT_USM_LEVEL_NONE
+    if constexpr (io != mem_inout::in) {
+      if (_source_attribute ==
+          dpct::detail::pointer_access_attribute::host_only) {
+        _q.memcpy(_source, _target, sizeof(target_t) * _ele_num).wait();
+      }
+    }
+#endif
+  }
+};
+} // namespace detail
+using out_mem_int64_int_t =
+    detail::mem_t<std::int64_t, int, detail::mem_inout::out>;
```

Comment:
For backward compatibility,  please don't export detail:: namespace class/function as public interface